observable_operator_test.go

```diff
@@ -323,6 +323,177 @@ func Test_Observable_BufferWithTimeOrCount(t *testing.T) {
+				seen = value
+			}
+		}
+		return nil
+	}))
+}
+
+func delayItem(ctx context.Context, item interface{}) (interface{}, error) {
+	entry := item.(struct {
+		name    string
+		msDelay int
+	})
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-time.After(time.Duration(entry.msDelay) * time.Millisecond):
+		return entry.name, nil
+	}
+}
+
+func Test_Observable_BufferWithTimeOrCount_ElapsedCapturesPartial(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+	defer cancel()
+
+	observable := Just([]struct {
+		name    string
+		msDelay int
+	}{
+		{"a", 5},
+		{"b", 1000},
+	})().Map(delayItem, WithContext(ctx))
+
+	buffers := observable.BufferWithTimeOrCount(WithDuration(time.Millisecond*10), 2, WithContext(ctx))
+
+	Assert(ctx, t, buffers, CustomPredicate(func(items []interface{}) error {
+		partialEncountered := false
+		for _, item := range items {
+			buffer := item.([]interface{})
+			if len(buffer) == 1 {
+				partialEncountered = true
+			}
+		}
+		if !partialEncountered {
+			return errors.New("at least one partial observation should have occurred")
+		}
+		return nil
+	}))
+}
+
+func Test_Observable_BufferWithTimeOrCount_EmitsCompleteBuffers(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+	defer cancel()
+
+	// straggler not emitted before context timeout
+	observable := Just([]struct {
+		name    string
+		msDelay int
+	}{
+		{"a", 1},
+		{"b", 1},
+		{"c", 1},
+		{"d", 1},
+		{"e", 1000},
+	})().Map(delayItem, WithContext(ctx))
+
+	buffers := observable.BufferWithTimeOrCount(WithDuration(time.Millisecond*10), 2, WithContext(ctx))
+
+	Assert(ctx, t, buffers, CustomPredicate(func(items []interface{}) error {
+		for _, item := range items {
+			buffer := item.([]interface{})
+			if len(buffer) != 2 {
+				return errors.New("items should be bundles of two")
+			}
+		}
+		return nil
+	}))
+}
+
+func Test_Observable_BufferWithTimeOrCount_CapturesStraggler(t *testing.T) {
+	defer goleak.VerifyNone(t)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond*100)
+	defer cancel()
+
+	// straggler will stay in buffer until buffer timeout
+	observable := Just([]struct {
+		name    string
+		msDelay int
+	}{
+		{"a", 1},
+		{"b", 1},
+		{"c", 1},
+		{"d", 1},
+		{"straggler", 1},
+	})().Map(delayItem, WithContext(ctx))
+
+	buffers := observable.BufferWithTimeOrCount(WithDuration(time.Millisecond*10), 2, WithContext(ctx))
+
+	Assert(ctx, t, buffers, CustomPredicate(func(items []interface{}) error {
+		seen := false
+		for _, item := range items {
+			buffer := item.([]interface{})
+			for _, b := range buffer {
+				if b.(string) == "straggler" {
+					seen = true
+				}
+			}
+		}
+		if !seen {
+			return errors.New("straggler item not emitted")
+		}
+		return nil
+	}))
+}
+
+func Test_Observable_BufferWithTimeOrCount_DoneEmitsStraggler(t *testing.T) {
```

Comment:
I think the overall logic with the flush boolean makes sense. Yet, this test looks to be flaky. I'm getting the following error locally:
```
=== RUN   Test_Observable_BufferWithTimeOrCount_DoneEmitsStraggler
    assert.go:220: 
        	Error Trace:	assert.go:220
        	            				observable_operator_test.go:480
        	Error:      	straggler item not emitted
        	Test:       	Test_Observable_BufferWithTimeOrCount_DoneEmitsStraggler
--- FAIL: Test_Observable_BufferWithTimeOrCount_DoneEmitsStraggler (0.01s)
```

Btw I was wondering why the CI was always okay after repeating its execution but I noticed it was running cached tests so I've merged this quick [PR](https://github.com/ReactiveX/RxGo/pull/307).

Perhaps as the `checkBuffer` closure keeps growing in complexity, it would be better to extract it as a function that would simplify writing unit tests, instead of having to rely on an overall flaky scenario based on time. We could keep a nominal test scenario of `BufferWithTimeOrCount` to test the overall execution + detailed tests on `checkBuffer` to check all the possible cases.

What do you think?

Reply:
There's a subtle collaboration w/ the `checkBuffer` closure and the for/select loop that calls it.  It seems to me that pulling `checkBuffer` out would make the function more complex since it's no longer a closure and it would cause the unit tests to be brittle white box unit tests.

With that said, I fixed the flakey tests.

Reply:
I'm still getting errors:
```
--- FAIL: Test_Observable_BufferWithTimeOrCount_CapturesStraggler (0.01s)
    assert.go:220: 
                Error Trace:    assert.go:220
                                                        observable_operator_test.go:443
                Error:          straggler item not emitted
                Test:           Test_Observable_BufferWithTimeOrCount_CapturesStraggler
```

Reply:
So, after a lot of hacky futzing w/ timing, the tests seem to reliably pass now.  As I was tinkering with the timing, I definitely considered your suggestion again.

As it currently stands, the amount of code in `observable_operator.go` is gigantic.  If I did do that bit of refactoring with `checkBuffer`, my preference would be to break up `observable_operator.go` into separate files, e.g. `observable_buffered.go`, and then perform the refactoring.  Would that work be worth it?

Thoughts?